Vertical navbar

fhlavac opened the pull request:
### Task
https://github.com/ManageIQ/manageiq-ui-classic/issues/5759

@Hyperkid123 @martinpovolny @epwinchell 
- Rewritten structure generation - made it recursive and moved it to navbar.rb
- Rewritten and divided React main_menu component into level components
- Created recursive PropTypes
- Created main-menu tests

![vertical-navbar](https://user-images.githubusercontent.com/50696716/61709300-31246080-ad4f-11e9-8113-6fc2ac652a0a.png)

fixes https://bugzilla.redhat.com/show_bug.cgi?id=1734661

Hyperkid123 commented:
Adding a bit of context.

Right now it still uses the same logic as current navigation. This PR changes migrates the haml layout to React and adds a new function that generates JSON fro which the menu is rendered.

In future PR we will move all the show/hide, click/hover and any other logic to the react component. Currently that is handled via pf JS which is triggering these state changes based on css classes (not id or any other unique identifiers)

I know that @epwinchell has some requests to change the current behavior. We would like to introduce them once we start moving the logic to React component. @epwinchell if you could create GH issue with the changes you would like to see that would be great.

Hyperkid123 commented on app/javascript/components/main-menu/main-menu.jsx:
```diff
@@ -0,0 +1,37 @@
+  0: props => <TopLevel level={level} {...props} />,
+  1: props => <SecondLevel level={level} {...props} />,
+  2: props => <ThirdLevel level={level} {...props} />,
+})[level];
```
I think we should add a fallback component if somebody passes level value other than 0 - 3

```jsx
const getLevelComponent = level => ({
  0: props => <TopLevel level={level} {...props} />,
  1: props => <SecondLevel level={level} {...props} />,
  2: props => <ThirdLevel level={level} {...props} />,
})[level] ||  props => <ThirdLevel level={level} {...props} />;
```

This will always return ThirdLevel component if you gave the function any level that it does not cover.

Hyperkid123 commented on app/javascript/components/main-menu/second-level.jsx:
```diff
@@ -0,0 +1,42 @@
+  items,
+  level,
+}) => (
+  <li className={`list-group-item ${items.length > 0 && 'tertiary-nav-item-pf'}`} data-target={`#menu-${id}`}>
```
This will add `false` to className if the condition is not met. We don't want that.

Hyperkid123 commented on app/javascript/components/main-menu/top-level.jsx:
```diff
@@ -0,0 +1,50 @@
+        </div>
+        <ul className="list-group">
+          {
+                items.map(props => <MenuItem key={props.id} level={level + 1} {...props} />)
```
Put this on single line please.

martinpovolny commented:
Marking this as "WIP" as we do not want it merged (even if it was finished) until the release is out.

Hyperkid123 commented on app/views/layouts/_header.html.haml:
```diff
@@ -15,7 +15,7 @@
         %a{:class => "nav-item-iconic drawer-pf-trigger-icon", :title => "{{vm.notificationsIndicatorTooltip}}", "ng-click" => "vm.toggleNotificationsList()"}
           %span.fa.fa-bell
-          %span{"ng-class" => "{'badge badge-pf-bordered': vm.newNotifications}"} 
+          %span{"ng-class" => "{'badge badge-pf-bordered': vm.newNotifications}"}
```
I don't think this file should be changed :wink:

Hyperkid123 commented:
@miq-bot add_label react
@miq-bot add_label hammer/no

epwinchell commented:
@miq-bot add_reviewer @epwinchell

epwinchell commented:
@fhlavac Please rebase

himdel commented on app/presenters/menu/manager.rb:
```diff
@@ -66,6 +66,51 @@ def section_id_string_to_symbol(section_id_string)
       valid_sections[section_id_string]
     end
 
+    def menu_structure(placement=:default)
```
spaces around `=` please :)

epwinchell commented:
> if you could create GH issue with the changes you would like to see that would be great.

https://bugzilla.redhat.com/show_bug.cgi?id=1734181

martinpovolny commented:
Ok, @fhlavac : please un"WIP" the PR.

@epwinchell : good to go?

himdel commented:
Is this new menu expected to support all the existing features?
(Except for clicking on sections.)

Specifically, I'm not seeing any code dealing with:

* calling `miqCheckForChanges` before leaving a form
* opening links in a new window
* opening modals (via sendDataWithRx)

himdel commented:
(This should really not be ignoring anything from `link_params`, but it only reads the `href` it seems.)

martinpovolny commented:
@himdel : I sort of expected something like that ;-)

Can we somehow capture the functionality that you pointed out in some tests? Sounds to me like 3 pretty important concepts and all the tests are green.

himdel commented:
Well, link params provided the whole `%a` attributes hash in the view. So, it's *probably* the only place to look at, but also painful to test.

Though, since we're moving to react, a new snapshot test could cover this.

epwinchell commented:
@martinpovolny yes, reviewed/approved

fhlavac commented:
@martinpovolny @himdel 

- Added support for missing types from link_params - "modal", "big_iframe" and "new_window".
(It would be nice to have data where these types occur to be able to really test it. I verified the functionality only by changing some items with "default" type to the other types mentioned.)

- Fixed calling miqCheckForChanges before leaving a form.

himdel commented:
Missing now:

* pinning - the pin icon is gone, until you click on the second level div

* sometimes the second level section gets stuck as selected 

![bad](https://user-images.githubusercontent.com/289743/64351149-56acb700-cfe9-11e9-848c-1096a34db51b.png)

* highlighting active section/item on the 2nd level doesn't seem to work  (1st and 3rd seem to work)

epwinchell commented:
> pinning - the pin icon is gone, until you click on the second level div

@himdel The pinning is going away either now or in a follow-up PR

himdel commented:
@fhlavac any progress on the second level stuck issue? Or am I the only one seeing this?

himdel commented:
@miq-bot add_label wip

fhlavac commented:
@himdel I am not sure I really understand the required behavior of highlighting and pinning. Maybe if you had a second for an explanation...

Hyperkid123 commented:
@fhlavac is this still WIP or do you think you are finished?

fhlavac commented:
@Hyperkid123 we need to discuss the pinning funcionality, I am not sure if it works correctly but @epwinchell says that this functionality is going away? Should I completely remove it now or leave it there and remove it in some future PR?

Hyperkid123 commented:
@fhlavac it is going away but first we need to move the logic (expanding,hovering, etc.) to this new component and stop calling the initialization function. That will be in future PR. I would leave it for now.

himdel commented:
Ideally, we should not be merging something that doesn't work (talking about pinning specifically here).
So, if we're leaving the pinning in for now, it should be fully functional.
(Unless fixing it would take significant time...)

Hyperkid123 commented:
@himdel by functional I believe you mean the missing icon? That should not be a problem.

himdel commented:
Last time I checked, nothing happened once you actually clicked on the icon. Unless it is fixed now.. :)

Hyperkid123 commented:
OK I will work with Filip to fix it ASAP.

fhlavac commented:
> Last time I checked, nothing happened once you actually clicked on the icon. Unless it is fixed now.. :)
> 
> (also, "highlighting active section/item on the 2nd level doesn't seem to work (1st and 3rd seem to work)")

- Now I tried clicking on the icon and I don't see any bug :/
- Highlighting section  on the 2nd level should be already fixed

himdel commented:
You're right, I missed the last commit, 2nd level is fixed,
and not seeing any pinning issues anymore either (not sure why, but hey, it works :)).

Will recheck the link_parms bits and merge :)

himdel commented:
Issues:

* `link_params` is now only called for `[:href]` but still does all the logic - clean up please?
* that logic is now spread all over the place instead of that 1 method:
  * `checkForChanges`  is handled separately and `main-menu/top-level.jsx`, `main-menu/second-level.jsx` and `main-menu/third-level.jsx` - the logic is copied in 3 places instead of shared
  * `big_iframe` is handled in main-menu/helpers.js
  * `modal` is handled in main-menu/helpers.js for the href, and in those 3 copies for `sendDataWithRx`
  * `new_window` is handled by the 3 copies only
* `javascript:void(0)` for a href won't actually work in React 16.9., right?

So...

I think this works as intended now :+1: 

I'm not merging *yet*, please keep the shared logic in one place, as opposed to 5 places in 3 copies ;).

fhlavac commented:
@miq-bot add_label wip

martinpovolny commented:
@fhlavac, @Hyperkid123, @himdel : Guys, great work in this PR!

I am looking forward for this and believe that you will keep the quality high.

We are ahead of time with this, no need to rush.

fhlavac commented:
- shared logic related to type of item moved to helpers.js
- `javascript:void(0)` removed
- as @himdel said, link_params will be fixed in https://github.com/ManageIQ/manageiq-ui-classic/pull/6197

himdel commented on app/javascript/components/main-menu/helpers.js:
```diff
@@ -2,11 +2,13 @@ export const getHrefByType = (type, href, id) => (
     big_iframe: `/dashboard/iframe?id=${id}`,
     // eslint-disable-next-line no-script-url
-    modal: 'javascript:void(0);',
+    modal: '#',
```
`javascript:void(0)` is problematic for react,
`#` is problematic for hash router,
*nothing* used to be problematic for IE, but might just work for us, testing :)

himdel commented:
Ok, so...

`DoNav('#')` - triggers a state change when on `#foo`, to `#`
`DoNav('')` - triggers a full reload, ends on `#` always

But, 

`modal: '#'` - doesn't trigger a state change at all, even when on `#foo` (surprising, not sure we should rely on this)
`modal: 'https://google.com/'` - goes to google as expected
`modal: ''` - (together with a change so that the `|| href` doesn't trigger) - triggers a full page reload
`modal: undefined` - (same change ^) - works, looks correct in chrome and firefox


So... I think `undefined` may be the way :)

(I was using this diff:)

```diff
--- a/app/javascript/components/main-menu/helpers.js
+++ b/app/javascript/components/main-menu/helpers.js
@@ -1,10 +1,13 @@
-export const getHrefByType = (type, href, id) => (
-  {
-    big_iframe: `/dashboard/iframe?id=${id}`,
-    // eslint-disable-next-line no-script-url
-    modal: '#',
+export const getHrefByType = (type, href, id) => {
+  switch (type) {
+    case 'big_iframe':
+      return `/dashboard/iframe?id=${id}`;
+    case 'modal':
+      return undefined;
+    default:
+      return href;
   }
-)[type] || href;
+};
 ```

himdel commented:
Test:

* [ ] a modal item should not trigger checkforchanges
* [ ] opening a modal works, doesn't change router state
* [ ] checkforchanges result is respected

himdel commented on app/javascript/spec/main-menu/__snapshots__/main-menu.spec.js.snap:
```diff
@@ -191,7 +191,7 @@ exports[`Main menu test should render correctly 1`] = `
               <li
                 className="active list-group-item secondary-nav-item-pf"
-                data-target="#menu-vi"
+                data-target="menu-vi"
```
This is wrong... https://getbootstrap.com/docs/3.3/javascript/

All the examples in bootstrap docs use data-target with the `#`.

Or is there any patternfly code treating data-target differently?

Hyperkid123 replied:
@himdel that is a typo.

fhlavac replied:
Sorry, # is back in targets now :)

himdel commented:
> checkforchanges result is respected

Still "ignored" - now we never navigate anywhere when in a form.

@fhlavac @Hyperkid123 I don't think you can make this work *without* using proper click events.

---

The problem right now is that mousedown triggers the modal, if you click cancel, it returns false, so cancels the mousedown event.
**But** if you click OK, it returns true, so the mousedown event is allowed to happen. But, a click only happens after a mousedown+mouseup combo, and the mouseup event never happens because the mouse is in a different place after confirming the modal.

Also, using mousedown instead of click means you can actually show the about modal without meaning to. 

---

Can you please describe the issue with just using click?

Hyperkid123 commented:
@himdel the problem is that the onClick is for some reason overridden by some jQuery/pf/boostrap code. We cannot add onClick event on react component with the classNames of the menu. `.pf-nav-*` I think.

himdel commented:
@Hyperkid123 Aaaah, good, so this is solvable :). Most likely this is not patternfly or jquery - I'm almost sure it's bootstrap JS.

Which means none of the patternfly classes should be a problem,
and it will be the set of classes described in https://getbootstrap.com/docs/3.4/components/#navbar-default,
and https://getbootstrap.com/docs/3.4/javascript/#dropdowns

Essentially, all those `data-target` and `data-toggle` attributes need to go away, as well as any `dropdown` class, and `navbar-collapse` class.

If those classes are associated with any styles, we can duplicate the styles,
no point in hacking around a component that doesn't work when we can fix it :).

Hyperkid123 commented:
@himdel I am almost certain about some styling being associated with those classes. @fhlavac will try removing the data attributes.

(And i though this will be easy task)

himdel commented:
:+1: Let me know if you need any help :) I think I still remember some gory details of bootstrap js :)

himdel commented:
So.. I was looking at [patternfly js](https://github.com/himdel/patternfly/tree/master/src/js), and you're right, it's probably involved too..

@fhlavac it may help to look at https://github.com/himdel/patternfly/blob/master/src/js/patternfly-functions-vertical-nav.js to know the relevant classes.

But, there may be another alternative - I don't think we could just remove bootstrap js, so anything from bootstrap might still have to be fix, but as far as patternfly goes, we don't have to require all the javascript, we can pick. (More importantly, we can require all but that one file.)

himdel commented:
> we can require all but that one file.

```diff
diff --git a/app/javascript/packs/globals.js b/app/javascript/packs/globals.js
index 340deb2db..2b5fb3a77 100644
--- a/app/javascript/packs/globals.js
+++ b/app/javascript/packs/globals.js
@@ -18,9 +18,23 @@ require('jquery-ui/ui/widgets/sortable');
 require('jquery-ujs');
 require('jquery.observe_field');
 require('patternfly-bootstrap-treeview');
-require('patternfly/dist/js/patternfly.min');
 require('jquery.hotkeys');
 
+// all of patternfly js except for vertical navigation (patternfly-functions-vertical-nav.js)
+// list from https://github.com/patternfly/patternfly/blob/master/Gruntfile.js#L62-L85
+require('patternfly/dist/js/patternfly-settings.js');
+require('patternfly/dist/js/patternfly-functions-base.js')
+require('patternfly/dist/js/patternfly-functions-list.js')
+require('patternfly/dist/js/patternfly-functions-sidebar.js')
+require('patternfly/dist/js/patternfly-functions-popovers.js')
+require('patternfly/dist/js/patternfly-functions-data-tables.js')
+require('patternfly/dist/js/patternfly-functions-navigation.js')
+require('patternfly/dist/js/patternfly-functions-count-chars.js')
+require('patternfly/dist/js/patternfly-functions-colors.js')
+require('patternfly/dist/js/patternfly-functions-charts.js')
+require('patternfly/dist/js/patternfly-functions-fixed-heights.js')
+require('patternfly/dist/js/patternfly-functions-tree-grid.js')
+
 window.angular = require('angular');
 require('angular-ui-bootstrap');
 require('angular-gettext');
```

fhlavac commented:
CheckForChanges should now be respected.
- changed onMouseDown to onClick handling
- replaced PF list-group-item class with custom class menu-list-group-item
- updated tests

epwinchell commented:
@fhlavac We're missing some styling  around the "collapse" class. I'll try to figure it out and get back to you.
<img width="192" alt="Screen Shot 2019-10-29 at 12 37 15 PM" src="https://user-images.githubusercontent.com/1287144/67789386-4396f000-fa4a-11e9-80d6-9bb8796f548c.png">

epwinchell commented:
@miq-bot add_label wip

epwinchell commented:
@fhlavac This might be part of the problem. The "secondary-visible-pf" class is being applied

Master:
<img width="1314" alt="Screen Shot 2019-10-29 at 4 58 39 PM" src="https://user-images.githubusercontent.com/1287144/67808736-9d5ce180-fa6d-11e9-8a49-a384eb0b4a8b.png">

This branch:
<img width="1301" alt="Screen Shot 2019-10-29 at 4 58 26 PM" src="https://user-images.githubusercontent.com/1287144/67808737-9d5ce180-fa6d-11e9-8671-9994b726e424.png">

epwinchell commented:
@fhlavac Replacing "menu-list-group-item" with the standard "list-group-item" seems to solve the problem.

Hyperkid123 commented:
> @fhlavac Replacing "menu-list-group-item" with the standard "list-group-item" seems to solve the problem.

@epwinchell  That is something we cannot do. The `list-group-item` class is causing the issues with event handlers. That is why it was renamed.

fhlavac commented:
> @fhlavac We're missing some styling around the "collapse" class. I'll try to figure it out and get back to you.
> <img alt="Screen Shot 2019-10-29 at 12 37 15 PM" width="192" src="https://user-images.githubusercontent.com/1287144/67789386-4396f000-fa4a-11e9-80d6-9bb8796f548c.png">

Thank you, it should be fixed now :)

epwinchell commented:
@fhlavac Everything looks good now. Great job!

martinpovolny commented:
@himdel, @Hyperkid123 ?

himdel commented:
I'm planning to test the click changes tomorrow...

himdel commented:
My checklist... (review in progress...)

* [x] reflects changes to those files since Jun
* [ ] click
* [ ] miqCheckForChanges
* [ ] modals
* [ ] https://github.com/ManageIQ/manageiq-ui-classic/pull/5617
* [ ] ids for menu items (not seeing one in top-level for top-level *items*)

himdel commented:
So, @fhlavac we're really close .. everything we discussed previously is working as it should :+1: :)

Only 2 things missing:

* every menu item (1st, 2nd or 3rd level) must have an `<li ... id="menu_item_....">` as of https://github.com/ManageIQ/manageiq-ui-classic/pull/5063
  * this PR does that for 3rd level only
  * 1st level is only achievable via custom menu items, but it shoud work for them too - see below

* the changes from #5617 - it looks like the current code assumes that 1st level is always a section:

(with the custom button added as described in #5617)

master:

```
<li class="list-group-item" id="menu_item_custom_i1">
  <a href="/dashboard/iframe?id=custom_i1" onclick="return miqCheckForChanges();">
```

this PR:

```
<li class="menu-list-group-item secondary-nav-item-pf" data-target="#menu-custom_i1">
  <a href="/dashboard/iframe?id=custom_i1" target="_self" class="top-level">
```

(the rest of the markup seems the same +- whitespace)

And hovering on master:
![before](https://user-images.githubusercontent.com/289743/68028060-cf5b8700-fcab-11e9-85f9-6d1fcdc8a5de.png)

With this PR:
![after](https://user-images.githubusercontent.com/289743/68028068-d2ef0e00-fcab-11e9-9ea4-7a87bd16e996.png)

(note the extra `>`)

skateman commented:
The CSS can be simplified, the JSON conversion can be refactored, but in a future PR, I'm good with this. Clicked through the UI a little and it all worked, even the custom menu.

himdel commented:
Confirming custom menus are working :)

But now everything has an id, not only menu items.
@fhlavac can you please make sure only menu items get ids? Or, at least change the section ids to use the `menu_section_` prefix instead of `menu_item_`?

fhlavac commented:
Update:
- items have `menu_item_ `prefixes
- sections have `menu_section_` prefixes
- removed probably redundant `data-target`s at top and second level

himdel commented:
Merging when green :)

skateman commented:
ssh, it's green :wink:

himdel commented on app/javascript/packs/globals.js:
```diff
@@ -18,9 +18,23 @@ require('jquery-ui/ui/widgets/sortable');
-require('patternfly/dist/js/patternfly.min');
 require('jquery.hotkeys');
 
+// all of patternfly js except for vertical navigation (patternfly-functions-vertical-nav.js)
```
so, this bit didn't really work,
because `require('angular-patternfly')` requires the whole thing anyway.

Fixing in https://github.com/ManageIQ/manageiq-ui-classic/pull/6449,
and removing the call to `setupVerticalNavigation` which no longer works without `patternfly-functions-vertical-nav.js`.
